osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -120,44 +118,111 @@ private DbEncryptedSecret createEncrypted(final String deviceIdentification, fin
         }
     }
 
-    public Optional<TypedSecret> retrieveSecret(final String deviceIdentification, final SecretType secretType) {
-        final Date now = new Date();
-        final Long secretId = this.secretRepository.findIdOfValidMostRecent(deviceIdentification, secretType.name(),
-                now);
-        if (secretId == null) {
+    private Optional<TypedSecret> retrieveActiveSecret(final String deviceIdentification,
+            final SecretType secretType) {
+        final Optional<DbEncryptedSecret> encryptedSecret = this.getSingleDbEncryptedSecret(deviceIdentification,
+                secretType, SecretStatus.ACTIVE);
+        return encryptedSecret.isPresent() ? Optional.of(this.getTypedSecret(encryptedSecret.get())) : Optional.empty();
+
+    }
+
+    private Optional<DbEncryptedSecret> getSingleDbEncryptedSecret(final String deviceIdentification,
+            final SecretType secretType, final SecretStatus secretStatus) {
+        final List<DbEncryptedSecret> secretsList = this.secretRepository.findSecrets(deviceIdentification, secretType,
+                secretStatus);
+        if (secretsList.size() == 0) {
             return Optional.empty();
+        } else if (secretsList.size() > 1) {
+            throw new IllegalStateException(); //TODO set proper msg
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1135)